### ISSUE TITLE
FR: Osara Win Installation should backup original reaper-kb.ini

### ISSUE THREAD

**Issue by X-Raym** [jcsteh/osara#309]
Hi,

I was trying to install Osara but when I re-open REAPER (v6.13 winx64), all my scripts where unregistered from the action list... And all the associated keyboard shortcut. No Backup.

Did I do something wrong ? Is there a way to restore it ?

Thank your assistance,

**Comment by X-Raym** [comment]
Ok I see the issue. It is on the install process, there is a switch to "Replace existing key map" in Osara install.

Is there any  backup made if we choose this option ?

reaper-kb.file are extremly lightweight, please consider adding a backup... I lost one hundred of custom keyboard shortcuts and I'm kinda stuck.

**Comment by X-Raym** [comment]
Also, such sensitive switch may not be switch oned by default, really, it is very easy to miss.

**Comment by ScottChesworth** [comment]
NO backup = ouch. Sorry man. We have the replace key map option checked by default because the vast majority of OSARA users are using that map as ground zero.

Would it be possible to add backing up reaper-kb.ini to the Windows installer, @jcsteh? I think Robbie's already done this over on the Mac side.

@X-Raym, I wonder whether system restore would recover your previous reaper-kb-ini?

**Comment by X-Raym** [comment]
@ScottChesworth Thanks for confirming.

No system backup available in this folder by default. All i have is a several years old backup, which may restore some of my custom actions but for the rest, it is just gone...

**Comment by RDMurray** [comment]
@X-Raym  You could try checking the previous versions tab in the properties dialog for the folder.

I do think we should make a backup if replace keymap is checked by default.  It might also be a good idea to make it a yes/no question like the mac installer.  Is that possible with NSIS?  It can be easy to miss that checkbox.

**Comment by X-Raym** [comment]
> @X-Raym You could try checking the previous versions tab in the properties dialog for the folder.
This wasn't activated for this folder, no previous version available. I have to restore manually.

**Comment by jcsteh** [comment]
I definitely agree we should create a backup.

I'm not a fan of the yes/no question for the Windows installer. It would make the installer even more dialog heavy than it is already. Also, while we don't currently do this, I believe it's possible to have the value of that option preserved across installs; i.e. if you uncheck it once, it could stay unchecked. If we move to a yes/no dialog, the benefit of that is minimal.

**Comment by X-Raym** [comment]
@jcsteh as long as there is a backup file I'm okay for the installer to stay as it is, as I understand your audience : the less popup and messages, the easier it will be. I should have evaluated more carefully when installing but I had only limited time and I didn't expect an extension to remove my keyboard sortcuts and custom action so I didn't took time to read. 
But with a backup file, it would have been just a little annoyance.

**Comment by X-Raym** [comment]
Several hours of debugging later and I'm still not even close to have restore my set up cause now my reaper-kb file is [buggy](https://forum.cockos.com/showthread.php?t=242294).
This backup feature should be considered high priority. Osara installation is way to risky for now.

**Comment by ScottChesworth** [comment]
I sympathize with the loss of data. Adding backup functionality can and will be done, but with respect, reading installer prompts and backup of files containing hundreds of customizations should also be considered high priority. As you say, the file is only lightweight, dropping a safety copy in Dropbox or similar would only take a couple of seconds.

**Comment by X-Raym** [comment]
@ScottChesworth I restart the installation to try to understand where I failed : for my defense, the installation text just displays "Replace existing key map with Osara key map". Key maps can be contextual to a GUI window for eg, virtual keyboard has his own keymap that we can customize in a dedicated file.  There is even a special dedicated KeyMaps folder in REAPER folders. There isn't one keymaps in reaper, there is several.
But the text doesn't explicitly says "Replace the current user main REAPER key map (containing both keymaps AND custom actions AND registered scripts)".. I had no way to figure out that this warning was about this main file and not just an extra .ini files added by the extension, for which a replacement could be proposed if it has been cusotmized by the user at a previous installation. The reaper-kb file is way more than a keymap file has it is used to also registered macros. That's why it isn't in the Keymaps folder.

But be reassured, I learned the backup lesson. The hard way.

Anyway, thanks @jcsteh for having taking this episode into account. User errors happen, and it has to be considered in any human centered design approach.

Best !

**Comment by kara-louise** [comment]
@ScottChesworth  I also ran in to the same issue as @X-Raym  a while ago, learning the hard way that the file mentioned in the installation is the ini file, not the key map in the key maps folder.
 I also had no idea that the ini file contains macros, or  what  must be references to installed scripts including ones downloaded from ReaPack. None of them would appear in the action list even tho all the files were still there. I initially thought this was an issue with ReaPack, not Osara as I didn’t know what the ini file contains, or which file was meant until I asked for help on the rwp list.
I had to reinstall everything from all the repos I’m subscribed to including @X-Raym‘s before they would show  up in the actions list again.
